src/RoombaBlobDetector.cpp

```diff
@@ -0,0 +1,310 @@
+                                 settings_.hsv_slice_s_max,
+                                 settings_.hsv_slice_v_max),
+                      range_mask);
+    cv::cuda::bitwise_or(dst, range_mask, dst);
+    // Lower red slice
+    cv_utils::inRange(hsv_image,
+                      cv::Scalar(settings_.hsv_slice_h_red2_min,
+                                 settings_.hsv_slice_s_min,
+                                 settings_.hsv_slice_v_min),
+                      cv::Scalar(settings_.hsv_slice_h_red2_max,
+                                 settings_.hsv_slice_s_max,
+                                 settings_.hsv_slice_v_max),
+                      range_mask);
+    cv::cuda::bitwise_or(dst, range_mask, dst);
+
+    cv::Mat structuring_element = cv::getStructuringElement(
+            cv::MORPH_RECT,
+            cv::Size(settings_.morphology_size, settings_.morphology_size));
+    cv::Ptr<cv::cuda::Filter> morphology = cv::cuda::createMorphologyFilter(
+            cv::MORPH_OPEN,
+            CV_8UC1,
+            structuring_element,
+            cv::Point(-1, -1),
+            settings_.morphology_iterations);
+    morphology->apply(dst, dst);
+
+    ROS_ASSERT(dst.channels() == 1);
+}
+
+void RoombaBlobDetector::boundMask(const cv::cuda::GpuMat& mask,
+                                   std::vector<cv::RotatedRect>& boundRect)
+{
+    cv::Mat mask_cpu;
+    mask.download(mask_cpu);
+
+    std::vector<std::vector<cv::Point>> contours;
+
+    //////////////////////////////////////////////////////////////////////////
+    /// Find bounding contours of blobs in mask
+    //////////////////////////////////////////////////////////////////////////
+
+    // findContours does not exist for the gpu
+    cv::findContours(mask_cpu,
+                     contours,
+                     CV_RETR_EXTERNAL,
+                     CV_CHAIN_APPROX_SIMPLE);
+
+    if (settings_.debug_contours) {
+        cv::Mat contour_image = cv::Mat::zeros(mask.rows, mask.cols, CV_8UC3);
+
+        for (const auto& contour : contours) {
+            cv_utils::drawContour(contour_image,
+                                  contour,
+                                  cv::Scalar(255, 255, 255));
+        }
+
+        const cv_bridge::CvImage cv_image {
+            std_msgs::Header(),
+            sensor_msgs::image_encodings::RGB8,
+            contour_image
+        };
+
+        debug_contours_pub_.publish(cv_image.toImageMsg());
+    }
+
+    //////////////////////////////////////////////////////////////////////////
+    /// Get orientation for each found contour
+    //////////////////////////////////////////////////////////////////////////
+
+    boundRect.resize(0); // Clear the vector
+    cv::RotatedRect rect;
+    for (unsigned int i = 0; i < contours.size(); i++) {
+        cv::Moments moments = cv::moments(contours[i]);
+
+        // throw out blobs that are too small or too large
+        if (moments.m00 < 2000 || moments.m00 > 15000) continue;
```

Comment:
Should this range be a parameter or is it very unlikely to ever change?

Reply:
Fixed